iterative/resource_task.go

```diff
@@ -138,6 +138,26 @@ func resourceTask() *schema.Resource {
 							Optional: true,
 							Default:  "",
 						},
+						"container": {
+							Type:     schema.TypeString,
+							ForceNew: true,
+							Optional: true,
+							Default:  "",
+						},
+						"container_path": {
+							Type:     schema.TypeString,
+							ForceNew: true,
+							Optional: true,
+							Default:  "",
+						},
+						"container_opts": {
+							Type:     schema.TypeMap,
+							ForceNew: true,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
```

Comment:
I'd consider taking a single [`rclone` connection string](https://rclone.org/docs/#connection-strings) instead of several separate arguments, as stated in the following reviews:
* https://github.com/iterative/terraform-provider-iterative/pull/660#pullrequestreview-1107692940
* https://github.com/iterative/terraform-provider-iterative/pull/651#pullrequestreview-1095672221

Reply:
What are the benefits of doing that?

Reply:
What @casperdcl said.

Reply:
> _would be worth checking if the rclone library can expose the parsed components_

```go
package main

import (
        "log"

        "github.com/rclone/rclone/fs"
        "github.com/rclone/rclone/lib/bucket"
        _ "github.com/rclone/rclone/backend/s3"
)

const remote = ":s3,provider=AWS,region=us-west-1,access_key_id=ASI...,secret_access_key=xMO...,session_token=IQo...:example-bucket/prefix"

func main() {
        _, _, path, config, err := fs.ParseRemote(remote)

        if err != nil {
                log.Fatal(err)
        }

        container, containerPath := bucket.Split(path)
        log.Println(container, containerPath)

        for key, value := range config {
                log.Println(key, value)
        }
}
```

Reply:
> _because we actually use rclone internally_

Except for Kubernetes, which only takes the identifier of a StorageClass or a PersistentVolume, but we can make an exception as usual.

Reply:
Even shorter, no connection strings involved:

#### Before
```hcl
container = "container-name"
container_path = "subdirectory"
container_opts = {
  account = "storage-account-name"
  key = "storage-account-key"
}
```
#### After
```
storage = "storage-account-name/container-name/subdirectory" # key retrieved automatically
```

Reply:
Ok, so if I understand correctly, the arguments for switching to a connection string type thing are:

1. we're already using rclone internally
2. connection strings are well documented

I'd argue with that.

1. what we use internally should have no bearing on what UX we expose to the users
2. the connection string format proposed (`storage = "storage-account-name/container-name/subdirectory"`) is not an rclone connection string, so we won't be able to rely on rclone's documentation, we'll be documenting a new connection string format (oh, wait, weren't we trying to avoid that?).

And additionally, the proposed format (`storage = "storage-account-name/container-name/subdirectory"`) is terrible:
1. It's positional, so it's more difficult to read.
2. It's also a format that will be difficult to extend ~~if~~when we need to add a config option to it (e.g. support for S3's `checksum` mode).
3. This format will also make it difficult to use different bucket authentication options (e.g. Azure's `sas_url`s).

Maybe a decent middle-ground would be combining `container` and `container_path` into `container="container_name/container/path"`? And leaving `container_opts` as they are?

Reply:
"Note that https://github.com/iterative/terraform-provider-iterative/pull/687#discussion_r992907995 is a separate proposal I posted after realizing that there was no need to expose options at all. Indeed, it's not a valid `rclone` connection string._

Reply:
> _what we use internally should have no bearing on what UX we expose to the users_

Err... to some extent 😅

Reply:
The current implementation of `container_opts` is mainly used by azure, since it allows us to specify either the storage account credentials (`account` and `key`) or the SAS url (`sas_url`).

Reply:
So for a task running on AWS we'd only need to specify `container=<name>[/<path>]`.

Reply:
> _The current implementation of container_opts is mainly used by azure, since it allows us to specify either the storage account credentials (account and key) or the SAS url (sas_url)._

Couldn't we take care of those details by default? Except for the storage account name, of course.

Reply:
We can, I'm working on a change to retrieve the storage account key if it's not specified. But the need to specify the storage account name still remains. And we can't say for certain we won't need to add other options in the future

Reply:
Ah, now I remember why I abandoned the idea of looking up storage account credentials before - to be able to look up the storage account `key` or `sas_url`, we need not only the name of the storage account, but also the resource group it belongs to. So instead of the user specifying `account` and `key`, they'd be specifying `account` and `resouce_group` - this hardly makes it any easier.

Reply:
I'm playing catch-up a bit here. can you help me out with a basic example of what `main.tf` might look like for our big three providers (AWS, GCP, Azure)?


In general, I feel like I agree more with @tasdomas just because the mechanism is rclone doesn't mean that should be exposed to the user... I don't think that a rclone connection string would be any more convenient for the average user, if anything I feel it would be more inconvenient.

I would go back to some of my original comments: [this one](https://github.com/iterative/terraform-provider-iterative/issues/299#issuecomment-982012371) and [down from here](https://github.com/iterative/terraform-provider-iterative/issues/299#issuecomment-985369985)

Reply:
All examples expect an empty `task` directory to exist.

# AWS
```
terraform {
  required_providers { iterative = { source = "github.com/iterative/iterative" } }
}

provider "aws" {
  region = "us-east-1"
}

resource "aws_s3_bucket" "bucket" {
  bucket = "test-tpi-bucket"
}

resource "aws_s3_bucket_acl" "bucket_acl" {
  bucket = aws_s3_bucket.bucket.id
  acl    = "private"
}

provider "iterative" {}

resource "iterative_task" "example" {
  cloud      = "aws" # or any of: gcp, az, k8s
  region     = "us-east-1"
  machine    = "m"   # medium. Or any of: l, xl, m+k80, xl+v100, ...
  spot       = -1     # auto-price. Default -1 to disable, or >0 for hourly USD limit
  disk_size  = -1    # GB. Default -1 for automatic
  storage {
    workdir = "./task"       # default blank (don't upload)
    output  = "results" # default blank (don't download). Relative to workdir
    container = "${aws_s3_bucket.bucket.bucket}/tpi-run"
  }
  script = <<-END
    #!/bin/bash

    # create output directory if needed
    mkdir -p results
    # read last result (in case of spot/preemptible instance recovery)
    if test -f results/epoch.txt; then EPOCH="$(cat results/epoch.txt)"; fi
    EPOCH=$${EPOCH:-1}  # start from 1 if last result not found

    echo "(re)starting training loop from $EPOCH up to 1337 epochs"
    for epoch in $(seq $EPOCH 1337); do
      sleep 1
      echo "$epoch" | tee results/epoch.txt
    done
  END
}
```

# Azure
```
terraform {
  required_providers { iterative = { source = "github.com/iterative/iterative" } }
}

provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "tpi_storage_rg" {
  name     = "tpi-storage-resources"
  location = "East US"
}

resource "azurerm_storage_account" "tpi_storage_account" {
  name                     = "tpistorage124154"
  resource_group_name      = azurerm_resource_group.tpi_storage_rg.name
  location                 = azurerm_resource_group.tpi_storage_rg.location
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_storage_container" "bucket" {
  name                  = "tpistoragetpi"
  storage_account_name  = azurerm_storage_account.tpi_storage_account.name
  container_access_type = "private"
}

provider "iterative" {}

resource "iterative_task" "example" {
  cloud      = "aws" # or any of: gcp, az, k8s
  region     = "us-east-1"
  machine    = "m"   # medium. Or any of: l, xl, m+k80, xl+v100, ...
  spot       = -1     # auto-price. Default -1 to disable, or >0 for hourly USD limit
  disk_size  = -1    # GB. Default -1 for automatic
  storage {
    workdir = "./task"       # default blank (don't upload)
    output  = "results" # default blank (don't download). Relative to workdir
    container = "${azurerm_storage_container.bucket.name}/tpi-run"
  }
  script = <<-END
    #!/bin/bash

    # create output directory if needed
    mkdir -p results
    # read last result (in case of spot/preemptible instance recovery)
    if test -f results/epoch.txt; then EPOCH="$(cat results/epoch.txt)"; fi
    EPOCH=$${EPOCH:-1}  # start from 1 if last result not found

    echo "(re)starting training loop from $EPOCH up to 1337 epochs"
    for epoch in $(seq $EPOCH 1337); do
      sleep 1
      echo "$epoch" | tee results/epoch.txt
    done
  END
}
```

# GCP

```
terraform {
  required_providers { iterative = { source = "github.com/iterative/iterative" } }
}

provider "google" {
  region = "us-east1-b"
}

resource "google_storage_bucket" "bucket" {
  name = "test-tpi-bucket"
  storage_class = "Standard"
  location = "US"
  project = "cml-gcp-test"
  force_destroy = true
}

provider "iterative" {}

resource "iterative_task" "example" {
  cloud      = "gcp" # or any of: gcp, az, k8s
  region     = "us-east1-b"
  machine    = "m"   # medium. Or any of: l, xl, m+k80, xl+v100, ...
  spot       = -1     # auto-price. Default -1 to disable, or >0 for hourly USD limit
  disk_size  = -1    # GB. Default -1 for automatic
  storage {
    workdir = "./task"       # default blank (don't upload)
    output  = "results" # default blank (don't download). Relative to workdir
    container = "${google_storage_bucket.bucket.name}/tpi-run"
  }
  script = <<-END
    #!/bin/bash

    # create output directory if needed
    mkdir -p results
    # read last result (in case of spot/preemptible instance recovery)
    if test -f results/epoch.txt; then EPOCH="$(cat results/epoch.txt)"; fi
    EPOCH=$${EPOCH:-1}  # start from 1 if last result not found

    echo "(re)starting training loop from $EPOCH up to 1337 epochs"
    for epoch in $(seq $EPOCH 1337); do
      sleep 1
      echo "$epoch" | tee results/epoch.txt
    done
  END
}
```

Reply:
IDK with your current set of just container[_opts] I think this is totally fine. Definitely clearer (IMO) than a single key that would be a rclone connection string.

Reply:
After https://github.com/iterative/terraform-provider-iterative/pull/687/commits/158568b5b03608b5fbe59745265867572549fe2e this conversation can be deemed as resolved.

I'd still mildly argue that `container_opts` shouldn't be needed anyway and, if it ever does, those options can be moved to a connection string, but that's probably in the realms of bikeshedding.